src/Dapper.AOT.Analyzers/Internal/Roslyn/TypeSymbolExtensions.cs

```diff
@@ -0,0 +1,171 @@
+{
+    public static bool InNamespace(this ITypeSymbol? typeSymbol, string @namespace)
+    {
+        if (typeSymbol is null) return false;
+        return typeSymbol.ContainingNamespace?.ToDisplayString() == @namespace;
+    }
+
+    public static string? GetTypeDisplayName(this ITypeSymbol? typeSymbol)
+    {
+        if (typeSymbol is null) return null;
+        if (typeSymbol.IsAnonymousType == true) return "object?";
+        return "global::" + typeSymbol.ToDisplayString();
+    }
+    public static string? GetContainingTypeDisplayName(this ITypeSymbol? typeSymbol, int typeArgIndex = 0)
+    {
+        var containingTypeSymbol = typeSymbol.GetContainingTypeSymbol(typeArgIndex);
+        return containingTypeSymbol.GetTypeDisplayName();
+    }
+    public static ITypeSymbol? GetContainingTypeSymbol(this ITypeSymbol? typeSymbol, int typeArgIndex = 0)
+    {
+        if (typeSymbol is null) return null;
+        if (typeSymbol is INamedTypeSymbol namedTypeSymbol)
+        {
+            var typeArguments = namedTypeSymbol.TypeArguments;
+            if (typeArgIndex >= typeArguments.Length) return null;
+            return typeArguments[typeArgIndex];
+        }
+        if (typeSymbol is IArrayTypeSymbol arrayTypeSymbol)
+        {
+            return arrayTypeSymbol.ElementType;
+        }
+
+        return null;
+    }
+
+    /// <returns>
+    /// True, if passed <param name="typeSymbol"/> represents <see cref="ImmutableArray{T}"/>.
+    /// False otherwise
+    /// </returns>
+    public static bool IsImmutableArray(this ITypeSymbol? typeSymbol)
+    {
+        if (typeSymbol is null) return false;
+        if (typeSymbol is not INamedTypeSymbol namedTypeSymbol) return false;
+
+        return namedTypeSymbol.TypeKind == TypeKind.Struct
+            && namedTypeSymbol.InNamespace("System.Collections.Immutable")
+            && namedTypeSymbol.Arity == 1
+            && namedTypeSymbol.Name == "ImmutableArray";
+    }
+
+    /// <returns>
+    /// True, if passed <param name="typeSymbol"/> represents array.
+    /// False otherwise
+    /// </returns>
+    public static bool IsArray(this ITypeSymbol? typeSymbol)
+    {
+        if (typeSymbol is null) return false;
+        if (typeSymbol is not IArrayTypeSymbol arrayTypeSymbol) return false;
+
+        return arrayTypeSymbol.TypeKind == TypeKind.Array
+            && arrayTypeSymbol.BaseType?.InNamespace("System") == true
+            && arrayTypeSymbol.BaseType?.Name == "Array";
+    }
+
+    /// <returns>
+    /// True, if passed <param name="typeSymbol"/> represents <see cref="List{T}"/>.
+    /// False otherwise
+    /// </returns>
+    public static bool IsList(this ITypeSymbol? typeSymbol)
+    {
+        if (typeSymbol is null) return false;
+        if (typeSymbol is not INamedTypeSymbol namedTypeSymbol) return false;
+
+        return namedTypeSymbol.TypeKind == TypeKind.Class
+            && namedTypeSymbol.InNamespace("System.Collections.Generic")
```

Comment:
since this is allocatey, maybe check the namespace *last* ? (in all the places where this is used)

Reply:
Sorry, I dont understand your suggestion.
What I can do to avoid string allocation is to precreate some known NamespaceSymbols (as statics) and just use them instead. Is that better?